diff -Nur -x '*.orig' -x '*~' unstable/bzrlib/__init__.py unstable.new/bzrlib/__init__.py
--- unstable/bzrlib/__init__.py	2009-04-30 20:47:26.000000000 +0200
+++ unstable.new/bzrlib/__init__.py	2009-04-30 20:48:47.000000000 +0200
@@ -53,7 +53,7 @@
 version_info = (1, 14, 0, 'final', 0)
 
 # API compatibility version: bzrlib is currently API compatible with 1.13.
-api_minimum_version = (1, 14, 0)
+api_minimum_version = (1, 13, 0)
 
 
 def _format_version_tuple(version_info):
